[WIP] Identify working tests not requiring setup of a node

lucafavatella opened the pull request:
In scope of https://www.pivotaltracker.com/story/show/164548714

This PR aims at identifying the tests that are known to be working, so that those can be scripted for regression testing and the test coverage can be assessed.

I consider identifying state of tests requiring setup of a node out of scope for this PR: I plan to address those (i.e. `tx_primops_eqc` and `txs_oldnew_eqc`) in following PR.

TODO:
* [x] Assess state of any remaining test not requiring setup of a node:
  * [x] `txs_hardfork_eqc`
* [ ] ~Port history of [oracles tests](https://github.com/aeternity/aeternity/tree/d9521c3863395d3a74a4f38f06cc788095537611/apps/aeoracle/eqc) over to this repo.~
  * ~See https://github.com/Quviq/epoch-eqc/pull/10~

Commit:
Move non-working system tests to separate folder

They rely on min tx fee being statically 1.

Commit:
Move non-working aeutils tests to separate folder

Commit:
Avoid property test folder names same as app names

Plugin `rebar3-eqc-plugin` [calls
`rebar_utils:update_code/2`](https://github.com/kellymclaughlin/rebar3-eqc-plugin/blob/8d5d189cfb43d479186f8b064122f76e6fa3ccec/src/rebar3_eqc.erl#L306)
that [calls
`code:del_path/1`](https://github.com/erlang/rebar3/blob/df53ad18ff5330660bd102d3a4071cb0962c8dda/src/rebar_utils.erl#L789)
with atom argument, and [that `code:del_path/1` with atom
argument](http://erlang.org/documentation/doc-9.1/lib/kernel-5.4/doc/html/code.html#del_path-1)
is relaxed in matching items to remove from the path (e.g. matching
`extras/aesophia` in addition to `lib/aesophia/ebin`):

```
del_path(NameOrDir) -> ...
...
Name = atom()
...

Deletes a directory from the code path. The argument can be an atom
Name, in which case the directory with the name
`.../Name[-Vsn][/ebin]` is deleted from the code path.  ...
```

Commit:
Move non-working aesophia test to separate folder

Commit:
Move non-working aens test to separate folder

Not working for usage of `eqc_mocking` (I checked the code path is
correct).  Symptom:
```
  Exception:
    {{undef,
	[{aec_governance, contributors_messages_hash, [], []},
	 {aens_eqc, init, 1,
```

Commit:
Move non-working aeoracles test to separate folder

Symptom:
```
  Exception:
    {{undef,
	[{aec_chain_state, new_from_persistance,
```

Commit:
Ease debugging test failure for missing contract file

Commit:
Refresh PoW test following module rename

Commit:
Refresh sync test following key/micro blocks and module rename

Commit:
Move non-working sync test to separate folder

Symptom:
```
aec_sync_eqc:start_peers() ->
  exit({noproc, {gen_server, call, [aec_keys, peer_pubkey]}}).
```

Commit:
Comment reference to nonexistent contract

Commit:
Fix id typo (`aser_id`) in test txs_glue test

Commit:
Update calls to id module

Commit:
Move tests requiring setup of a node to separate folder

Test module `txs_oldnew_eqc` requires `txs_eqc` so attempting to run
that manually becomes slightly more difficult.  Though having only
tests not requiring setup of a node in the `aecore_eqc` folder eases
scripting running those.

Commit:
Teach txs test to find hard fork accounts files

Symptom:
```
txs_eqc:multi_mine(0, 10) ->
  !!! {exception,
	 {'EXIT',
	    {{minerva_accounts_file_missing,
		"../..//Users/ae/dev/ae/aeternity/data.nonode@nohost/aecore/.minerva/accounts_test.json"},
```

Commit:
Teach txs glue test to find hard fork accounts files

Symptom:
```
txs_glue_eqc:multi_mine(0, 10) ->
  exit({{{minerva_accounts_file_missing,
	    "/Users/ae/dev/ae/aeternity/data.nonode@nohost/aecore/.minerva/accounts_test.json"},
```

Commit:
Prevent tests from leaking hard fork settings

I noticed test `txs_glue_eqc` was failing for unrecognized protocol
version `3` when run after `txs_eqc`.

I am not asserting that `aecore` app is not loaded before running
property in order to be permissive with current (tag 1.2.0) behaviour
of rebar3 eqc plugin:
https://github.com/kellymclaughlin/rebar3-eqc-plugin/blob/8d5d189cfb43d479186f8b064122f76e6fa3ccec/src/rebar3_eqc.erl#L40
https://github.com/erlang/rebar3/blob/3.9.1/src/rebar_utils.erl#L775-L798

Commit:
Move test requiring features not merged to master

Generalized accoount meta transactions are not merged in master yet so
strictly speaking they are not working yet.

Test module `txs_meta_eqc` requires `txs_eqc` so attempting to run
that manually becomes slightly more difficult.  Though having only
working tests in the `aecore_eqc` folder eases scripting running
those.

ThomasArts commented on aecore_eqc/txs_eqc.erl:
```diff
@@ -1434,6 +1437,21 @@ prop_txs(Fork) ->
                                            {total, Total == 0 orelse equals(Total, ?PatronAmount - FeeTotal)}]))))))))
     end)))).
 
+load_app_if_unloaded(App) ->
```
far too verbose.... this is test code. Just load it no optimizations

lucafavatella replied:
Ok - I will make load less verbose.

lucafavatella replied:
Addressed in 29164569fc8043b5e34e50797d123d765c314cea

ThomasArts commented on aecore_eqc/txs_eqc.erl:
```diff
@@ -1893,7 +1911,9 @@ gen_ttl() ->
 gen_contract() ->
-    elements(["identity", "authorize_nonce"]).
+    elements([ "identity"
+             %% , "authorize_nonce" TODO Restore `authorize_nonce.aes` alongside `identity.aes` in https://github.com/aeternity/aeternity/tree/master/test/contracts
```
should now be in PR

lucafavatella replied:
I see https://github.com/aeternity/aeternity/pull/2315

lucafavatella replied:
So I need to adapt this PR where authorize_nonce used.

ThomasArts commented on aecore_eqc/txs_glue_eqc.erl:
```diff
@@ -224,13 +224,13 @@ spend_args(#{accounts := Accounts, height := Height} = S) ->
                    {name, Name} -> {name, Name}
                end,
-               #{sender_id => aser_id:create(account, Sender#account.key),  %% The sender is asserted to never be a name.
+               #{sender_id => aeser_id:create(account, Sender#account.key),  %% The sender is asserted to never be a name.
```
good catch.. this could never have worked....

ThomasArts reviewed (changes requested):
Cannot take the complete PR in, but will fix a few things, such as to make more rotten properties work again.

Commit:
Prevent test from leaking mocking setup

I noticed test `txs_hardfork_eqc` was failing when run after
`txs_glue_eqc`.

Commit:
Teach txs hardfork test to find hard fork accounts files

Commit:
Refresh txs_hardfork test to consider recent changes in txs_eqc re contract create

Symptom:
```
       {undef,
	  [{txs_hardfork_eqc, contract_create,
```

Commit:
Refresh txs_hardfork test to consider recent changes in txs_eqc re protocol

Symptom:
```
    {function_clause,
        [{txs_hardfork_eqc,aggregate_feats,
             [[],[{protocol,ns_claim,0}],{eqc_gen,#Fun<eqc_gen.37.80374633>}],
```

Commit:
Move non-working txs_hardfork test to separate folder

Symptom:
```
===== txs_hardfork_eqc:prop_txs
********************************************************************************************************
OK, passed 104 tests
...
10.7% {txs_hardfork_eqc,contract_create,5}
...
Failed! Commands not executed: contract_call
...
correct
36.6% false
...
 4.6% contract_create
...
contract_create
56.0% {error,illegal_vm_version}
31.2% ok
 5.7% {error,too_low_gas_price}
 4.3% {error,too_low_fee}
 2.1% {error,account_nonce_too_low}
 0.7% {error,account_nonce_too_high}

contract_call
...
===> The following properties failed: [prop_txs,prop_txs]
```

Commit:
Move non-working txs test to separate folder

Symptom:
```
===== txs_eqc:prop_txs1
********************************************************************************************************
OK, passed 104 tests
...
11.0% {txs_eqc,contract_create,5}
...
Failed! Commands not executed: contract_call
...
correct
38.7% false
...
 3.6% contract_create
...
contract_call_fun

...
contract_create
33.3% {compiler,2}
30.5% {error,illegal_vm_version}
16.7% {compiler,1}
12.4% ok
 4.3% {error,too_low_gas_price}
 2.1% {error,too_low_fee}
 0.4% {error,account_nonce_too_low}
 0.4% {error,account_nonce_too_high}

contract_call

...
===> The following properties failed: [prop_txs]
```

Commit:
Make tests setup re aecore app less verbose

Addresses comment in PR 6.

lucafavatella commented on README.md:
```diff
@@ -3,6 +3,6 @@
  * Check out this repo into e.g. `epoch/eqc`
  * Start a shell using rebar3 (from `epoch/`): `./rebar3 as test shell --apps=""`
- * Move into the directory containing the models you would like to run, e.g.: `1> cd("eqc/aeutils").`
+ * Move into the directory containing the models you would like to run, e.g.: `1> cd("eqc/aeutils_eqc").`
```
I see this doc rename included in https://github.com/Quviq/epoch-eqc/commit/e2b1542f77c1511c5fc400861877b0c26bd27193
Doc update proposed as https://github.com/Quviq/epoch-eqc/pull/9

lucafavatella commented:
Re 002d7c7
test `txs_eqc` is now fixed by https://github.com/Quviq/epoch-eqc/commit/8f45c985438e986d473c3a87af9464e3571b2b24

lucafavatella commented:
`txs_glue` test fix split as #11

lucafavatella commented:
`aec_pow` test fix split as #12

lucafavatella commented:
Mostly cherry-picked and merged, or being handled as other PRs e.g. https://github.com/Quviq/epoch-eqc/pull/26

So closing this.
